### ISSUE TITLE
It is slow with a lot of data

### ISSUE THREAD

**Issue by kaka-ruto** [akodkod/solid-queue-dashboard#2]
Thanks for the beautiful UI!

I have 250,000 finished jobs, and the first page for the solid queue dashboard could not load anything at all

And postgres maxes out it's CPU usage when trying to retrieve the data

<img width="1440" alt="Screenshot 2024-10-08 at 03 10 30" src="https://github.com/user-attachments/assets/6f2bcb4a-f9c5-45b0-9c95-a61a297de015">

<img width="389" alt="Screenshot 2024-10-08 at 03 14 14" src="https://github.com/user-attachments/assets/29effe04-a8e0-4ba1-9297-e89bf2a46819">

**Comment by akodkod** [comment]
Got you. Thank you for feedback. I'll try it on 250K records today.

**Comment by akodkod** [comment]
@kaka-ruto Apologies for the late reply. Could you try the new version? I’ve generated 250K dummy events, and everything worked flawlessly.

**Comment by zhuravel** [comment]
Also having this issue. I only have around 2.5 million jobs in MySQL and the page is loading more than 5 minutes.

These queries are slow:

> SELECT COUNT(*) FROM solid_queue_jobs WHERE solid_queue_jobs.finished_at IS NOT NULL AND solid_queue_jobs.id NOT IN (SELECT solid_queue_jobs.id FROM solid_queue_jobs LEFT OUTER JOIN solid_queue_failed_executions failed_execution ON failed_execution.job_id = solid_queue_jobs.id WHERE failed_execution.id IS NOT NULL) AND solid_queue_jobs.id NOT IN (SELECT solid_queue_jobs.id FROM solid_queue_jobs WHERE solid_queue_jobs.finished_at <= '2024-10-22 16:09:43' AND solid_queue_jobs.id NOT IN (SELECT solid_queue_jobs.id FROM solid_queue_jobs LEFT OUTER JOIN solid_queue_failed_executions failed_execution ON failed_execution.job_id = solid_queue_jobs.id WHERE failed_execution.id IS NOT NULL) AND solid_queue_jobs.active_job_id IN (SELECT solid_queue_jobs.active_job_id FROM solid_queue_jobs GROUP BY solid_queue_jobs.active_job_id HAVING (COUNT(*) > 1)) AND solid_queue_jobs.id NOT IN (SELECT MAX(id) FROM solid_queue_jobs GROUP BY solid_queue_jobs.active_job_id))

> SELECT COUNT(*) FROM solid_queue_jobs WHERE solid_queue_jobs.finished_at IS NOT NULL AND solid_queue_jobs.id NOT IN (SELECT solid_queue_jobs.id FROM solid_queue_jobs LEFT OUTER JOIN sol
id_queue_failed_executions failed_execution ON failed_execution.job_id = solid_queue_jobs.id WHERE failed_execution.id IS NOT NULL) AND solid_queue_jobs.id NOT IN (SELECT solid_queue_job
s.id FROM solid_queue_jobs WHERE solid_queue_jobs.finished_at <= '2024-10-22 16:09:43' AND solid_queue_jobs.id NOT IN (SELECT solid_queue_jobs.id FROM solid_queue_jobs LEFT OUTER JOIN so
lid_queue_failed_executions failed_execution ON failed_execution.job_id = solid_queue_jobs.id WHERE failed_execution.id IS NOT NULL) AND solid_queue_jobs.active_job_id IN (SELECT solid_q
ueue_jobs.active_job_id FROM solid_queue_jobs GROUP BY solid_queue_jobs.active_job_id HAVING (COUNT(*) > 1)) AND solid_queue_jobs.id NOT IN (SELECT MAX(id) FROM solid_queue_jobs GROUP BY sol
id_queue_jobs.active_job_id))

> SELECT COUNT(*) FROM solid_queue_jobs WHERE solid_queue_jobs.finished_at <= '2024-10-22 16:16:10' AND solid_queue_jobs.id NOT IN (SELECT solid_queue_jobs.id FROM solid_queue_jobs LEFT OUTER JOIN solid_queue_failed_executions failed_execution ON failed_execution.job_id = solid_queue_jobs.id WHERE failed_execution.id IS NOT NULL) AND solid_queue_jobs.active_job_id IN (SELECT solid_queue_jobs.active_job_id FROM solid_queue_jobs GROUP BY solid_queue_jobs.active_job_id HAVING (COUNT(*) > 1)) AND solid_queue_jobs.id NOT IN (SELECT MAX(id) FROM solid_queue_jobs GROUP BY solid_queue_jobs.active_job_id)

Would it be possible to make it configurable to not do these counts, or do counts with limit, like showing `10K+` instead of total count? I really like the UI and the work you’ve done on this gem but can’t use it because of performance.

**Comment by kaka-ruto** [comment]
Thanks for the update Kod! It's now much better. My stats are now 400k. The homepage took a while to load, about 7s, while the stats page completely failed to load, resulted in a timeout.

Personally I would love to see all numbers but as @zhuravel mentioned, perhaps we can start with a `10k+` display before we get a more performant way to show all numbers.

Thanks again!

**Comment by thisIsLoading** [comment]
i dont know if mysql supports this but postgres has materialized views. maybe we could have one that updates every x minutes.

i will be getting roughly 500,000 jobs per day once my app is done and i dont see this working out other than with some form of caching.

with this number of jobs per day, i dont think it needs to be up to the minute data.

or we get some kind of a counter cache into solid_queue, but i am not sure where to put it, given its table structure

**Comment by kaka-ruto** [comment]
Yeah, I wonder how the Mission Control gem does its counting, vis a vis this gem's? (looked at it a bit but not enough to draw conclusions). Anyone done?

**Comment by thisIsLoading** [comment]
when you hit a certain threshold they start to cheat a little bit:

![image](https://github.com/user-attachments/assets/b419ef75-743e-4db6-aaa2-01c2e99176fa)

**Comment by kushagra-03** [comment]
> Yeah, I wonder how the Mission Control gem does its counting, vis a vis this gem's? (looked at it a bit but not enough to draw conclusions). Anyone done?

Maybe this could be of help: https://github.com/fatkodima/fast_count

**Comment by thisismydesign** [comment]
https://github.com/rails/mission_control-jobs doesn't have this issue
